### PR TITLE
v0.4.0 -- Group View

### PR DESCRIPTION
**Pull request by sjgilbert**
- New Main view links to Join and Make
- Main shows all groups user is a member of
- New Group view shows groupID, creator and members
- Renamed ParseApplication to UnanimusApplication
- Touched up logic in Join activity
- Changed "members" in UnanimusGroup to array of Strings (usernames)

### CODE REVIEW

File: ParseStarterProject/src/com/parse/starter/GroupActivity.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+        //Setting the group name at top
+        TextView groupNameTextView = (TextView) findViewById(R.id.group_name);
+        groupNameTextView.setText("GROUP ID: " + groupName);
+
+        //Query for the group's data
+        ParseQuery<UnanimusGroup> query = ParseQuery.getQuery("UnanimusGroup");
+        query.include("members");
+        try{
+            group = query.get(groupName);
```
Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/GroupActivity.java:43]: This is the climax.

File: ParseStarterProject/src/com/parse/starter/LoginActivity.java [hunk 2]
```diff
@@ -40,7 +40,7 @@ private void login() {
         String username = usernameEditText.getText().toString().trim();
         String password = passwordEditText.getText().toString().trim();
 
-        if(username.length() != 0 && password.length() != 0) {
+        if(username.length() != 0 && password.length() != 0) {  //requires user and pass not empty
```
Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/LoginActivity.java]: ildf

Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/LoginActivity.java]: Instead, try this:

``` java
// ...
private void login() throws SomeSortOfException {
    String username = usernameEditText.getText().toString().trim();
    String password = passwordEditText.getText().toString().trim();

    if(minUserLen > username.length() || minPassLen > password.length()) {
        String errMsg = String.format("Usernames must have at least %d characters and passwords must have at least %d characters", minUserLen, minPassLen);
        throw new SomeSortOfException(errMsg);
    }

    final ProgressDialog wait = new ProgressDialog(LoginActivity.this);
    // ...
}
// ...
private final static int minUserLen = 6;
private final static int minPassLen = 4;
// ...
```

File: ParseStarterProject/src/com/parse/starter/GroupActivity.java (hunk 1 above)
Comment (sjgilbert) [inline ParseStarterProject/src/com/parse/starter/GroupActivity.java:43]: What do you mean the climax?

File: ParseStarterProject/src/com/parse/starter/LoginActivity.java (hunk 2 above)
Comment (sjgilbert) [inline ParseStarterProject/src/com/parse/starter/LoginActivity.java]: But the user won't be able to see the errMsg that way. (I think?)
Toasts are a way of flashing an error message to the user.  What about this?

``` java
boolean error = false;
StringBuilder errorMessage = new StringBuilder();
if (username.length() < minUserLen) {
        error = true;
        errorMessage.append("Username too short");
}
if (password.length() < minPassLen) {
        if (error) {
                errorMessage.append(", and ");
        }
        error = true;
        errorMessage.append("Password too short");
}
errorMessage.append(".");

if (error) {
        Toast.makeText(LoginActivity.this, errorMessage.toString(), Toast.LENGTH_LONG);
        return;
}
```

Comment (sjgilbert) [inline ParseStarterProject/src/com/parse/starter/LoginActivity.java]: I realize I've also gotta do this for the RegisterActivity.  Any thoughts on how to format the appended message so it displays the minimum number of chars?

Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/LoginActivity.java]: The problem is that login exits successfuly when it was not successful.  The parent needs to know that.

``` Java
// ...
public void onCreate(Bundle savedInstanceState) {
    // ...
    loginButton.setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            try { login(); } catch (SomeSortOfException e) {
                Toast.makeText(LoginActivity.this, e.getMessage());
                return;
} /* ... */ } /* ... */ } }
// ...
```

**Comment by Izzette** [comment]
What if the parse server responded with the minimum length of usernames and passwords.  That way we could change that value without users updating their version, but users who made accounts with the old rules would still be able to log in.

File: ParseStarterProject/src/com/parse/starter/LoginActivity.java (hunk 2 above)
Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/LoginActivity.java]: To stick along with your factory method idea, what about this:

``` java
// ...
private void login() throws SomeSortOfException {
    String username = usernameEditText.getText().toString().trim();
    String password = passwordEditText.getText().toString().trim();

    boolean isError = false;
    StringBuilder errMsg = new StringBuilder();

    if(minUserLen > username.length()) {
        isError = true;
        errMsg.append(String.format("Username must be at least %d characters in length", minUserLen);
    }

    if (minPassLen > password.length()) {
        if(isError) errMsg.append("\n"); else isError = true;
        errMsg.append(String.format("Password must be at least %d characters in length", minPassLen);
    }

    if (isError) throw new SomeSortOfException(errMsg.toString());

    final ProgressDialog wait = new ProgressDialog(LoginActivity.this);
    // ...
}
// ...
private final static int minUserLen = 6;
private final static int minPassLen = 4;
// ...
```

File: ParseStarterProject/src/com/parse/starter/GroupActivity.java (hunk 1 above)
Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/GroupActivity.java:43]: I mean this is where the magic happens.  Like moment of truth stuff.

**Comment by Izzette** [comment]
Why should members be Strings?

File: ParseStarterProject/src/com/parse/starter/GroupActivity.java (hunk 1 above)
Comment (sjgilbert) [inline ParseStarterProject/src/com/parse/starter/GroupActivity.java:43]: Yep :-)

**Comment by sjgilbert** [comment]
I changed members to be UnanimusUsers.  The only change I really had to make in the rest of the code was to remove a couple getUsername()s, and I had to make this cast to UnanimusUser:
```java
 try {
                    first = (UnanimusGroup) query.get(groupIDStr);
                    if(first.setMember((UnanimusUser) ParseUser.getCurrentUser())) {
                        first.save();
                        Toast.makeText(JoinGroupActivity.this,"Success!",Toast.LENGTH_LONG).show();
                        return;
                    }
```
That cast should be harmless, right?

**Comment by sjgilbert** [comment]
Changed members to ParseUsers.  I tried making UnanimusUser class and making members of that type, but there are some issues with subclassing ParseUser.

**Pull request merged**